### PR TITLE
Collect Index methods by purpose: rendering, constructors, setops...

### PR DESCRIPTION
**Pull request by jbrockmendel**
Zero changes to the contents of _any_ methods, just collecting them in reasonably-scoped sections.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
travis failure appears unrelated

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -1468,12 +1573,218 @@ def rename(self, name, inplace=False):
...
+            msg = "'value' must be a scalar, passed: {0}"
+            raise TypeError(msg.format(type(value).__name__))
+
+    @property
+    def _has_complex_internals(self):
+        # to disable groupby tricks in MultiIndex
+        return False
+
+    def _summary(self, name=None):
         """
```
Comment (jreback) [inline pandas/core/indexes/base.py]: rendering

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -1959,537 +2416,432 @@ def get_duplicates(self):
...
+                pass
 
-        elif isinstance(state, tuple):
+        return self._shallow_copy(values)
 
-            if len(state) == 2:
-                nd_state, own_state = state
-                data = np.empty(nd_state[1], dtype=nd_state[2])
-                np.ndarray.__setstate__(data, nd_state)
-                self.name = own_state[0]
```
Comment (jreback) [inline pandas/core/indexes/base.py]: indexing methods

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -1959,537 +2416,432 @@ def get_duplicates(self):
...
+        https://github.com/pandas-dev/pandas/issues/19764
         """
-        Return a new Index of the values set with the mask.
+        if self.is_object() or self.is_categorical():
+            return name in self
+        return False
 
-        See Also
-        --------
-        numpy.ndarray.putmask
```
Comment (jreback) [inline pandas/core/indexes/base.py]: reshaping methods

File: pandas/core/indexes/base.py [hunk 4]
```diff
@@ -3117,35 +3463,20 @@ def symmetric_difference(self, other, result_name=None):
...
+            other = Index(other, name=self.name)
+            result_name = self.name
+        else:
+            result_name = get_op_result_name(self, other)
+        return other, result_name
 
-        return self._shallow_copy(values)
+    # --------------------------------------------------------------------
 
     _index_shared_docs['get_loc'] = """
```
Comment (jreback) [inline pandas/core/indexes/base.py]: indexing

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: what method was this comment supposed to refer to?

File: pandas/core/indexes/base.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: what method was this comment supposed to refer to?

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jreback) [reply pandas/core/indexes/base.py]: oh this section, basically group all of the indexing (and indexer conversion methods)

File: pandas/core/indexes/datetimes.py [hunk 5]
```diff
@@ -805,6 +647,172 @@ def intersection(self, other):
             left_chunk = left.values[lslice]
             return self._shallow_copy(left_chunk)
 
+    # --------------------------------------------------------------------
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: can you annotate these

File: pandas/core/indexes/base.py [hunk 6]
```diff
@@ -709,6 +699,170 @@ def dtype_str(self):
...
+        copy : bool, default True
+            By default, astype always returns a newly allocated object.
+            If copy is set to False and internal requirements on dtype are
+            satisfied, the original data is used to create a new Index
+            or the original Index is returned.
+
+            .. versionadded:: 0.19.0
+        """
+
+    @Appender(_index_shared_docs['astype'])
```
Comment (jreback) [inline pandas/core/indexes/base.py]: maybe this goes in conversion

File: pandas/core/indexes/base.py [hunk 7]
```diff
@@ -709,6 +699,170 @@ def dtype_str(self):
...
+                raise ValueError(msg)
+            taken = algos.take(values,
+                               indices,
+                               allow_fill=allow_fill,
+                               fill_value=na_value)
+        else:
+            taken = values.take(indices)
+        return taken
+
+    # ops compat
```
Comment (jreback) [inline pandas/core/indexes/base.py]: prob can remove this comment

File: pandas/core/indexes/base.py [hunk 8]
```diff
@@ -1468,144 +1614,304 @@ def rename(self, name, inplace=False):
...
+        """
+        return self
+
+    def _convert_for_op(self, value):
+        """
+        Convert value to be insertable to ndarray.
+        """
+        return value
+
+    def _assert_can_do_op(self, value):
```
Comment (jreback) [inline pandas/core/indexes/base.py]: maybe to the Array section?

File: pandas/core/indexes/base.py [hunk 9]
```diff
@@ -1468,144 +1614,304 @@ def rename(self, name, inplace=False):
...
+        False
+        """
+        return self.is_unique and self.is_monotonic_decreasing
+
+    def is_lexsorted_for_tuple(self, tup):
+        return True
+
+    @cache_readonly
+    def is_unique(self):
+        """
```
Comment (jreback) [inline pandas/core/indexes/base.py]: maybe consider these 2 introspection to the unique section

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jreback) [reply pandas/core/indexes/base.py]: IOW also include ``__getitem__``

File: pandas/core/indexes/base.py (hunk 3 above)
Comment (jreback) [reply pandas/core/indexes/base.py]: append/concat section

File: pandas/core/indexes/base.py [hunk 10]
```diff
@@ -3117,35 +3466,21 @@ def symmetric_difference(self, other, result_name=None):
...
+    def _convert_can_do_setop(self, other):
+        if not isinstance(other, Index):
+            other = Index(other, name=self.name)
+            result_name = self.name
+        else:
+            result_name = get_op_result_name(self, other)
+        return other, result_name
 
-        return self._shallow_copy(values)
+    # --------------------------------------------------------------------
```
Comment (jreback) [inline pandas/core/indexes/base.py]: would move these near the ssection you currently have named indexer

**Review by jreback — CHANGES_REQUESTED** [review]
generally ok, some comments. ideally put the same sections in the subclasses, even if they are empty (could be followup if needed)

File: pandas/core/indexes/datetimes.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: These are end-of-section separators; the methods below are un-grouped.  I can collect all the groups at the top so there is only one un-annotated?

File: pandas/core/indexes/base.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: I considered that, thought that ndarray-like methods should take priority

File: pandas/core/indexes/base.py (hunk 8 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: this is only actually used for `fillna`, so maybe in the null-handling section?  its a weird fit

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: Not asking what section, asking what method.  On GH your comment is pointing towards `__setstate__`, which I really doubt you mean should go in the indexing section

File: pandas/core/indexes/base.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: Same, asking what method, not what section

File: pandas/core/indexes/base.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: sure

**Comment by jbrockmendel** [comment]
collected uncategorized methods in one block just above generated-methods.  implemented the subset of suggestions that didn't need clarification.

This is liable to become a rebasing nightmare if left open.  I think should be multi-pass process.

File: pandas/core/indexes/base.py (hunk 10 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks. if any comments you didn't address here, pls do in a followup.

**Pull request merged**